Expand internal architecture guides (#153026)
elastic/elasticsearch: docs/internal/DistributedArchitectureGuide.md

without knowing its total length up front. The body is instead broken into a sequence of length-prefixed chunks
terminated by a final zero-length chunk. Elasticsearch deals with this on both the request and the response side, and
the two are handled very differently.

On the request side, chunked request bodies are decoded transparently by Netty's stock `HttpRequestDecoder`
([`Netty4HttpServerTransport`](https://github.com/elastic/elasticsearch/blob/v9.3.0/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java)),
which parses chunked and `Content-Length`-based bodies into the same stream of `HttpContent` objects followed by a
`LastHttpContent`. Netty removes the chunk framing itself, so Elasticsearch only ever sees the decoded body delivered
incrementally. Elasticsearch does not implement its own chunk-framing logic for incoming requests.
[`Netty4HttpRequestBodyStream`](https://github.com/elastic/elasticsearch/blob/v9.3.0/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestBodyStream.java)
then exposes those chunks to the rest of the HTTP layer as an incremental `HttpBody.Stream` (see the Sync/Async IO

> discussion above).

Reviewer: I'd remove this part. I guess intention was to say "we follow HTTP spec for "transfer-encoding: chunked" and use Netty's implementation for chunk encoding/decoding". `Netty4HttpRequestBodyStream` is not related to chunk-encoding. "chunk" such an overloaded term, can be easily confused.

Author: Removed the last sentence related to `Netty4HttpRequestBodyStream`

The response side is more interesting, because it lets Elasticsearch produce a chunked response without ever
materializing the full response body in memory. A handler whose response implements
[`ChunkedToXContent`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContent.java)
returns an iterator of `ToXContent` chunks rather than a single fully-built document.
[`ChunkedRestResponseBodyPart`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBodyPart.java)
drives that iterator, serializing only as many chunks as are needed to keep the outbound network buffer full and
pausing once it fills up. This lets Elasticsearch stream arbitrarily large responses at roughly constant memory,

> instead of building the whole response as one large in-memory buffer before sending the first byte.

Reviewer: With one caveat, it's still possible to have a large response in memory unserialized. Chunked response provides a callback to produce next part, implementation might choose produce each part lazily or have it all upfront and use chunking to iterate over some sort of in memory collection. So chunking reduces serialized copy of the Java response object. Where size is bounded to iterator's pass size. In pathological cases it can be a collection with single gigabyte item. In most cases large collections of not large elements are good candidates for chunked response.

Author: I removed the claim of "roughly constant memory" to address the caveat you pointed out.

[`RestChunkedToXContentListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/rest/action/RestChunkedToXContentListener.java)
is the common integration point a REST handler uses to return a `ChunkedToXContent` response this way. The resulting

> `HttpContent`s are ultimately written out by Netty's stock

Reviewer: typo: stock -> stack

Author: Replaced stock with built-in to avoid confusion.

### ThreadPool

The [`ThreadPool`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java) class is the central registry for the node's shared named executors and the entry point for obtaining one. Rather than holding

> references to raw `ExecutorService`s, code asks the `ThreadPool` for a named executor and submits work to it.

Reviewer: The placeholder has legit questions (why many thread pools, whats the purpose). 

A couple of things worth adding to this section:

State the convention explicitly: ES doesn't spin up ad-hoc threads or executors — every Runnable runs on a designated pool obtained from ThreadPool. The main exceptions are third-party libraries, such as SDKs (AWS/GCP/Azure clients) that manage their own internal thread pools.

Why there are many pools: resource isolation (bulkheading) — a saturated pool for one workload can't starve another — plus per-workload sizing, back-pressure policy, and per-pool monitoring (`_cat/thread_pool`, node stats). A pool is not just a named executor; each has specific sizing/queue/monitoring rules.

Add a worked example — WRITE: fixed at allocatedProcessors threads with a large bounded queue `(max(allocatedProcessors * 750, 10000))`, tracked for execution time/utilization. It performs blocking disk IO via Lucene, which is exactly why it's a dedicated bounded pool. Contrast with Netty transport-worker threads, which must never block — this "which runnable may run where" distinction (blocking vs non-blocking) is a core rule and could reference the Netty section.

Author: edited the thread pool section with this info

- [`RefCountingListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java)
  handles a dynamic or unknown number of tasks, or cases where you only need to await all of them without collecting results.
- [`SubscribableListener<T>`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java)
  lets multiple independent callers `addListener` to the same result so they are all notified on completion, and its `andThen`
  supports sequential CPS chaining of steps.
- [`ThreadedActionListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java)
  wraps a delegate and dispatches its completion onto a specified `Executor`. This is the standard way to fork a
  listener's continuation off the Netty event-loop thread (see the [Networking](#networking) / Netty discussion above and
  the [Performance](#performance) note below) so that expensive follow-up work does not run on an I/O thread.

> #### Long-Lived Listeners

Reviewer: I'd call them Pub/Sub vs One-Shot listeners.

### ThreadPool

The [`ThreadPool`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java) class

> is a central registry that owns most executors in a node and is the entry point for getting one. Rather than holding

Reviewer 2: Nit:

```suggestion
The ThreadPool class is the central registry for the node's shared named executors and the entry point for obtaining one.
```

Author: applied

- A **FIXED** pool has a static number of threads, backed by a *bounded* or *unbounded* queue. When the queue fills up, new tasks are
  *rejected immediately* (surfacing as an `EsRejectedExecutionException`). This gives predictable resource consumption and
  fast back-pressure under load, so it is used for the hot data paths such as `SEARCH` and `WRITE`.
- A **SCALING** pool has a thread count that grows on demand between a min and a max, backed by an *unbounded* queue, and
  its idle threads expire after a keep-alive period. This suits bursty, infrequent work such as `GENERIC`, `MANAGEMENT`,
  and snapshots.

#### Built-in pools

The built-in pools are declared as string constants in
[`ThreadPool.Names`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L77).
Plugins register additional pools at startup by overriding
[`Plugin::getExecutorBuilders(Settings)`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/plugins/Plugin.java#L304).

> `CLUSTER_COORDINATION` carries all cluster state updates. It is sized to a **single thread by design** so that

Reviewer 2: How about something like "`CLUSTER_COORDINATION` handles cluster-state publication, master elections, and cluster membership updates." to clarify different thread pools are used for cluster state update computation and application?

Author: Sounds good. I removed the cluster state update computation and application part from this line.

cluster state updates, master elections and cluster membership updates are serialized and correctness is guaranteed (see the [Cluster Coordination](#cluster-coordination) section).
`GENERIC` scales up to
[somewhere between 128 and 512 threads](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/DefaultBuiltInExecutorBuilders.java#L33)
depending on the number of processors, and is used throughout the codebase. That high limit makes it tempting to
block on a `Future`. Doing so risks deadlock if that future's completion

> itself depends on other work queued on `GENERIC`.

Reviewer 2: > That high limit makes it tempting to block on a Future. Doing so risks deadlock if that future's completion itself depends on other work queued on GENERIC.

Nit: I would remove this. The blocking-on-Future caveat is explained in the next section, so it feels a bit early + confusing here.

Author: Removed

#### `AbstractRunnable` and `ActionRunnable`

Submitting a plain `Runnable` via `execute(...)` works, but the task gets none of the structured handling ES relies on.

> There is no way to distinguish *rejection* (the queue was full) from a normal run, and any exception thrown from `run()`

Reviewer 2: How about:
```suggestion
There is no built-in way to distinguish and handle *rejection* (when the queue is full) separately from a normal run exception
```

Author: Reads better, so I applied it

- [`RefCountingListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java)
  handles a dynamic or unknown number of tasks, or cases where you only need to await all of them without collecting results.
- [`SubscribableListener<T>`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java)
  lets multiple independent callers `addListener` to the same result so they are all notified on completion, and its `andThen`
  supports sequential CPS chaining of steps.
- [`ThreadedActionListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java)
  wraps a delegate and dispatches its completion onto a specified `Executor`. This is the standard way to fork a
  listener's continuation off the Netty event-loop thread (see the [Networking](#networking) / Netty discussion above and
  the [Performance](#performance) note below) so that expensive follow-up work does not run on an I/O thread.

#### Long-Lived Listeners

> Distinct from the one-shot `ActionListener` are the *long-lived* listeners. Registered once with a service, they live for

Reviewer 2: I would avoid emphasizing *long-lived* here. Something like "event-based, usually long-lived" also feels more precise, long-lived is not really a vetted term.

Author: Modified

### ThreadPool

> The [`ThreadPool`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java)

Reviewer 2: Nitty nit: The `ThreadPool` is -> The `ThreadPool` class is

> is the central registry that owns every executor in a node and is the entry point for getting one. Rather than holding

Reviewer 2: Might have been an overstatement from my end 😇  `ThreadPool` is the central registry for the named shared executors on a node, but it does not literally own every executor. Most do go through it, but there are exceptions. For example, `ClusterApplierService` creates and owns its own dedicated single-thread executor

Author: I updated the text so that it "owns most executors".

references to raw `ExecutorService`s, code asks the `ThreadPool` for a named executor and submits work to it:

```java
threadPool.executor(ThreadPool.Names.SEARCH).execute(runnable);
threadPool.generic().execute(runnable);
threadPool.schedule(runnable, delay, executor);
threadPool.scheduleWithFixedDelay(runnable, interval, executor);
```

#### Two pool types

Each pool has a [`ThreadPoolType`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L159):

> - A **FIXED** pool has a static number of threads backed by a *bounded* queue. When the queue fills up, new tasks are

Reviewer 2: ```suggestion
- A **FIXED** pool has a static number of threads and can be backed by a bounded or unbounded queue . If the queue fills up, new tasks are
```
Double checked and some FIXED pools have unbounded queues:
https://github.com/elastic/elasticsearch/blob/b134cfba5141a49526f740803c6b35ce234ddee4/server/src/main/java/org/elasticsearch/threadpool/DefaultBuiltInExecutorBuilders.java#L190-L193

- A **SCALING** pool has a thread count that grows on demand between a min and a max, backed by an *unbounded* queue, and
  its idle threads expire after a keep-alive period. This suits bursty, infrequent work such as `GENERIC`, `MANAGEMENT`,
  and snapshots.

#### Built-in pools

The built-in pools are declared as string constants in
[`ThreadPool.Names`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L77).
Plugins register additional pools at startup by overriding
[`Plugin::getExecutorBuilders(Settings)`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/plugins/Plugin.java#L304).
Notable pools and their purpose:

- `GENERIC` (scaling) is the catch-all for recovery, housekeeping, and ad-hoc work.
- `SEARCH` (fixed) runs user search queries on the data nodes.
- `SEARCH_COORDINATION` (fixed) coordinates the search fan-out across shards.
- `WRITE` (fixed) handles indexing, bulk, and delete operations.
- `CLUSTER_COORDINATION` carries all cluster-state updates. It is sized to a **single thread by design** so that

>   cluster-state application is serialized and correctness is guaranteed (see the [Cluster Coordination](#cluster-coordination) section).

Reviewer 2: Not cluster state _application_, but updates (also master elections, and cluster membership): https://github.com/elastic/elasticsearch/blob/b134cfba5141a49526f740803c6b35ce234ddee4/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L105-L111

Author: Refactored to "...so that
  cluster state updates, master elections and cluster membership updates are serialized..."

- `MANAGEMENT` (scaling) serves health, monitoring, and admin read APIs.
- `FLUSH` / `REFRESH` / `MERGE` (scaling) run the corresponding shard maintenance operations, described under
  [Segments, Refresh, and Flush](#segments-refresh-and-flush) and [Segment Merges](#segment-merges).

#### `AbstractRunnable` and `ActionRunnable`

Submitting a plain `Runnable` works, but the executor cannot then tell the task that it was *rejected* rather than run,

> and any exception thrown from `run()` is swallowed by the worker thread. ES therefore prefers two richer abstractions:

Reviewer 2: "swallowed by the worker thread" seems a bit too strong. The failure may still surface via execute(...), but indeed plain `Runnables` do not get the structured `onFailure()` /`onRejection()` handling that `AbstractRunnable` provides.

Author: I refactored this a little bit to explain the semantics between execute and submit.

tokens, tracing IDs, custom request headers, the security context, and other transient objects that should flow with the
work but not be serialized over the wire. This context is preserved automatically across thread-pool hops. The
[`EsThreadPoolExecutor`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/common/util/concurrent/EsThreadPoolExecutor.java#L174)
wraps every submitted task in a `ContextPreservingRunnable` at submission time, and the worker thread restores the
captured context before `doRun()` and cleans it up afterwards. As a result a request that begins with a user's auth
token and fans out across search, fetch, and coordination threads carries that token on every hop with no manual
threading of context through the code.

### ActionListener

See the [Javadocs for `ActionListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionListener.java)

`ActionListener<T>` is the primary async callback interface in Elasticsearch, a form of *continuation-passing style*.

> Instead of computing and returning a value, an asynchronous operation is handed "what to do next" (the listener) as an

Reviewer 2: This is slightly overstating the "always asynchronous" model. Listeners can either complete synchronously on the calling thread or another thread, but there is no enforcement that they have to be asynchronous

Author: I think the text is referring to the async operation that happens before the listener is invoked. If that operation is synchronous, we shouldn't need an action listener and can execute the listener right after the target operation returns in a procedural fashion.

I'll refactor this so that it flows better.

argument and invokes it when the result is ready, and the caller passes its continuation into the callee rather than blocking
until the callee finishes. This keeps threads free, lets operations compose without waiting, and, because listeners are
also invoked from thread-pool executors, naturally carries the `ThreadContext` (see [ThreadPool](#threadpool)) across
async hops. The contract is a two-path pair of methods:

```java
void onResponse(T response); // success path
void onFailure(Exception e);  // failure path
```

The `ActionListener` Javadoc explains why ES deliberately avoids `java.util.concurrent.CompletableFuture` and similar
mechanisms in production code. They can achieve the same goals, but they permit *blocking* while waiting for a result,
almost never appropriate where threads are a precious resource, and they can *catch an `Error`*, delaying (or preventing)
the JVM exit that should follow such a fatal condition. `ActionListener` makes those misuses impossible.

#### Composition

> You almost never subclass `ActionListener` directly. Instead you compose delegates using its static factory methods, for

Reviewer 3: > You almost never subclass `ActionListener` directly.

assuming by subclass we mean anonymous classes (like `new ActionListener<>() { ... }`

and even if your statement is true that we don't need to reach for it frequently

as a reader i read this more as "we don't use this frequently so you shouldn't reach for it"

which doesn't strike me as right, especially for listeners with a bit more code, just consider exception handling

```suggestion
You don't frequently need to subclass `ActionListener` directly, but when you do, consider error handling in `onResponse`. Instead you compose delegates using its static factory methods, for
```

Author: Refactored this a bit. Thanks!

- [`RefCountingListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java)
  handles a dynamic or unknown number of tasks, or cases where you only need to await all of them without collecting results.
- [`SubscribableListener<T>`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java)
  lets multiple independent callers `addListener` to the same result so they are all notified on completion, and its `andThen`
  supports sequential CPS chaining of steps.
- [`ThreadedActionListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java)
  wraps a delegate and dispatches its completion onto a specified `Executor`. This is the standard way to fork a
  listener's continuation off the Netty event-loop thread (see the [Networking](#networking) / Netty discussion above and
  the [Performance](#performance) note below) so that expensive follow-up work does not run on an I/O thread.

#### Lifecycle listeners

> Distinct from the one-shot `ActionListener` are the *lifecycle* listeners. Registered once with a service, they live for

Reviewer 2: I am not sure "Lifecycle listeners" is a vetted term 😄 How about "Long-Lived Listeners" or something better along those lines?

Author: Renamed to Long-lived listeners instead

> the lifetime of the node and follow the classical Observer pattern, in which a subject maintains a list of registered

Reviewer 2: ```suggestion
 live for the lifetime of the node (or until explicitly deregistered)
```

There are places where cluster state listeners get deregistered mid-runtime: https://github.com/elastic/elasticsearch/blob/899cfc85b878d68753f627bd1ff31e993b3b5d45/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java#L745-L746

Author: updated

listeners and notifies all of them on each event. There is no notion of completion, and the listener keeps receiving
events. The two most common are
[`ClusterStateListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/cluster/ClusterStateListener.java)
(fires on every cluster-state update) and
[`IndexEventListener`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java)
(per-index and per-shard lifecycle events).

### Chunk Encoding

Elasticsearch uses two completely separate serialization systems, and neither uses Java's built-in `Serializable`:

- The **transport binary format** is the node-to-node wire protocol, driven by the `Writeable` interface (see
  [Wire Serialization](#wire-serialization) below).
- **XContent** is the structured document format used by REST APIs and for on-disk cluster state, supporting JSON, YAML,
  SMILE, and CBOR via the `ToXContent` family of interfaces (see [XContent](#xcontent) below).

> 

Reviewer: Chunk Encoding and XContent sections are confusing. I'd suggest add one at a time and capture nuances.

Chunk Encoding is an [HTTP request and response concept](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Transfer-Encoding). We accept chunked requests and assemble chunks in the HTTP pipeline, it's handled by Netty HTTP decoder. We also support chunked responses, which is more interesting to document, since we can produce a streamed response without materializing entire response in memory (ChunkedToXContent)

Author: I removed references to XContent here and elaborated on Chunk Encoding a bit. Let me know what you think.

Author: Edits to X-Content and Chunk Encoding sections are now reverted. These sections can be revisited as part of a separate PR.

#### XContent

XContent is the structured, human-orientable serialization system. The
[`XContentType`](https://github.com/elastic/elasticsearch/blob/v9.3.0/libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentType.java)
enum enumerates the supported concrete formats, which are JSON, YAML, SMILE, and CBOR, plus versioned variants. The
abstraction exists so that most code does not hardcode a format. Cluster state is persisted on disk in **SMILE** (a
compact binary superset of JSON), REST APIs default to **JSON**, and a REST caller can request a different response

> format via `?format=`.

Reviewer: `XContent is the structured, human-orientable serialization system.` and `enum enumerates the supported concrete formats, which are JSON, YAML, SMILE, and CBOR`

Contradicts to itself. Binary formats are not human-orientable. The notion of XContent comes from the custom HTTP headers (x-content-type) (see second paragraph https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers). So XContent is a client oriented format. Everything that implements XContent and it's derivates can be serialized in the HTTP/REST layer.

I wouldn't mix cluster state persistence with XContent in the same paragraph (or even this section)

Author: I think I'll leave the XContent summary piece out for now and save it for a later change. Thanks for the correction!

**Writing** is done through two interfaces, both requiring
`XContentBuilder toXContent(XContentBuilder builder, Params params)`:

- [`ToXContentObject`](https://github.com/elastic/elasticsearch/blob/v9.3.0/libs/x-content/src/main/java/org/elasticsearch/xcontent/ToXContentObject.java)

>   wraps its output in `builder.startObject()` / `builder.endObject()`, and is the standard for top-level

Reviewer 2: The phrase "wrap its output" might be slightly confusing here?  Maybe something like "guarantees that the emitted content is a complete valid object rather than a fragment"

Author: Updated

  request/response objects.
- [`ToXContentFragment`](https://github.com/elastic/elasticsearch/blob/v9.3.0/libs/x-content/src/main/java/org/elasticsearch/xcontent/ToXContentFragment.java)
  writes bare key/value pairs without an enclosing object, for objects that are always embedded inside a larger object
  managed by the caller.

**Parsing** is preferentially done with a static `ConstructingObjectParser` declared on the class, which preserves
immutability by collecting constructor arguments before applying any setters:

```java

> private static final ConstructingObjectParser<Thing, Void> PARSER = new ConstructingObjectParser<>(

Reviewer 3: a note (it's easy to miss) about `ignoreUnknownFields` param would be nice

Author: Added the note

*deserialization constructor* taking a `StreamInput`, which keeps the type immutable and puts read and write logic side by
side:

```java
public MyClass(StreamInput in) throws IOException {
    this.count = in.readVInt();
    this.items = in.readCollectionAsList(Item::new);
}
```

[`StreamOutput`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java)
and [`StreamInput`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java)
are abstract `java.io.OutputStream` / `InputStream` subclasses that form the write and read sides of this layer. They are
*typed* streams that know how to encode ES's primitive and common types (`writeVInt`, `writeString`, collection helpers,
etc.), and, critically, each carries the negotiated `TransportVersion` for the connection, so `writeTo` implementations
and deserialization constructors can branch for backwards compatibility. The current BWC idiom gates a field on a named
transport-version feature:

```java
// writing

> if (out.getTransportVersion().supports(MY_FEATURE)) {

Reviewer: There is a `Versioning.md` in the same directory. Worth to reference or drop this section and redirect. That document is comprehensive.

Author: I'll leave the brief intro as-is since the link already exists.

  and snapshots.

#### Built-in pools

The built-in pools are declared as string constants in
[`ThreadPool.Names`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L77).
Plugins register additional pools at startup by overriding
[`Plugin::getExecutorBuilders(Settings)`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/plugins/Plugin.java#L304).
Notable pools and their purpose:

- `GENERIC` (scaling) is the catch-all for recovery, housekeeping, and ad-hoc work.
- `SEARCH` (fixed) runs user search queries on the data nodes.
- `SEARCH_COORDINATION` (fixed) coordinates the search fan-out across shards.
- `WRITE` (fixed) handles indexing, bulk, and delete operations.
- `CLUSTER_COORDINATION` carries all cluster-state updates. It is sized to a **single thread by design** so that
  cluster-state application is serialized and correctness is guaranteed (see the [Cluster Coordination](#cluster-coordination) section).
- `MANAGEMENT` (scaling) serves health, monitoring, and admin read APIs.

> - `FLUSH` / `REFRESH` / `MERGE` (scaling) perform shard maintenance operations.

Reviewer 4: There are sections for these operations: `Segments, Refresh, and Flush` and `Segment Merges` that can be linked from here.

argument and invokes it when the result is ready, and the caller passes its continuation into the callee rather than blocking
until the callee finishes. This keeps threads free, lets operations compose without waiting, and, because listeners are
also invoked from thread-pool executors, naturally carries the `ThreadContext` (see [ThreadPool](#threadpool)) across
async hops. The contract is a two-path pair of methods:

```java
void onResponse(T response); // success path
void onFailure(Exception e);  // failure path
```

The `ActionListener` Javadoc explains why ES deliberately avoids `java.util.concurrent.CompletableFuture` and similar
mechanisms in production code. They can achieve the same goals, but they permit *blocking* while waiting for a result,
almost never appropriate where threads are a precious resource, and they can *catch an `Error`*, delaying (or preventing)
the JVM exit that should follow such a fatal condition. `ActionListener` makes those misuses impossible.

#### Composition

You almost never subclass `ActionListener` directly. Instead you compose delegates using its static factory methods.

> - `wrap(onResponse, onFailure)` builds an inline listener from two lambdas.

Reviewer 4: Again this is not a full list and i would prefer to just say to look inside `ActionListener`.

from compute, keeps durability in an object store, and treats nodes as ephemeral workers. **Serverless** is the managed
*offering* built on that architecture. It abstracts away provisioning, autoscaling, and error handling, billing users
for shared resource usage rather than for statically provisioned servers. The public
[Elastic serverless architecture blog](https://www.elastic.co/blog/elastic-serverless-architecture) contains an
architecture diagram of the two-tier design, and the design is described in detail in the SoCC'25 paper
[*Elasticsearch on the cloud: a stateless architecture*](https://dl.acm.org/doi/10.1145/3772052.3772245).

Splitting the write path (`index` nodes) from the read path (`search` nodes) is what lets read and write capacity scale
horizontally and independently. Stateless is implemented as an Elasticsearch plugin (the
[`stateless`](https://github.com/elastic/elasticsearch/tree/main/x-pack/plugin/stateless) plugin) that overrides the
relevant stateful components while reusing most of the stateful concepts described elsewhere in this guide.

> The two tiers exchange all durable data through the object store:

Reviewer 2: Not "all", fresh commits can be fetched directly from index node to search node if they have not been uploaded yet.

I also think the write-path, read-path, and recovery details are a bit out of place in a section about roles specifically. Those topics probably deserve their own dedicated sections where they can be explained in more detail. How do you feel about removing them for now?

Author: Sounds good, I can remove the sub sections for now.

- **Write path.** On an `index` node the

>   [`IndexEngine`](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/stateless/src/main/java/org/elasticsearch/xpack/stateless/engine/IndexEngine.java)

Reviewer 2: Avoid linking `main` branch code in doc, as those references are expected to become obsolete or change. Instead, link a specific commit or tag

Author: Edited the other links as well for the stateless classes that were referenced.

- `index`: indexing nodes, which host primary shards and handle all write operations.
- `search`: search nodes, which host search-only replica shards and handle read operations.

It is worth distinguishing two terms that are often conflated. **Stateless** is the *architecture*. It separates storage
from compute, keeps durability in an object store, and treats nodes as ephemeral workers. **Serverless** is the managed

> *offering* built on that architecture. It runs on Kubernetes and abstracts away provisioning, autoscaling, and error

Reviewer 4: Kubernetes feels like a detail that can be skipped here.

This is the responsibility of the [ClusterApplierService] class, which runs on
a [single dedicated thread](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L159).

Note that, to save bandwidth, a follower normally receives only a *diff* against the state it last acknowledged rather
than the full serialized state (see the publication-side diff discussion above). The follower reconstructs the complete
`ClusterState` from that diff locally, and it is this reconstructed full state that is then exposed as the current state

> to the appliers and listeners below.

Reviewer 2: I think this only happens if the node local `lastSeenClusterState` matches the base state that the master used to compute the diff, so the latest publication. To double check

Author: I verified it and you're right. I updated this block to handle the caveat.

                ├── _state/
                │   └── state-<N>.st         # index metadata (settings, mappings)
                └── <shard-id>/              # e.g. 0, 1, 2 ...
                    ├── _state/
                    │   └── state-<N>.st     # shard state (routing, allocation ID)
                    ├── index/               # Lucene segment files + write.lock
                    │   ├── segments_N
                    │   └── ...              # see the Lucene File Layout table below
                    └── translog/
                        ├── translog-<N>.tlog # translog generation files
                        └── translog.ckp      # translog checkpoint (min/max seqno)
```

The [`node.lock`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/env/NodeEnvironment.java#L186)
file is a filesystem lock acquired by [NodeEnvironment] at startup, and it prevents two Elasticsearch processes from
sharing the same data directory (see the [Locking](#locking) section). The `_state/state-<N>.st` files are written
via `MetadataStateFormat` and hold the index metadata (at the index level) and shard routing/allocation state (at the

> shard level). The `-<N>` suffix is a monotonically increasing generation so a new state can be written and fsynced

Reviewer 2: I think the shard level `_state/state-<N>.st` might hold a little more narrow data than what is described here: see https://github.com/elastic/elasticsearch/blob/a59c182f9f7e9d1bf3d6eecbc0e44f24ff91d053/server/src/main/java/org/elasticsearch/index/shard/ShardStateMetadata.java#L37-L42

Author: Sounds good, updated the text to match reality.

before the old one is removed. The `index/` directory holds the Lucene segments and commit points (its contents are
detailed in the [Lucene File Layout](#lucene-file-layout) table). The `translog/` directory holds the durability log
described under [Translog](#translog), where `translog.ckp` is the checkpoint tracking the current generation and its
sequence-number range.

The runtime object graph that owns these files is a containment hierarchy rooted at the shard:
`IndexShard -> Engine (InternalEngine) -> { IndexWriter, DirectoryReader, Store -> Directory }`. The `IndexWriter`
writes documents and drives flushes and merges, the `DirectoryReader` opens segments for search, and the `Store`
wraps the Lucene `Directory` for raw file I/O. All of the Lucene extension points ES plugs in (the versioned codec,
the wrapped `TieredMergePolicy`, soft deletes, and the index deletion policy) are assembled in

> [`InternalEngine#getIndexWriterConfig()`](https://github.com/elastic/elasticsearch/blob/v9.3.0/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java),

Reviewer 2: Not all, if my understanding is correct (eg:  query cache and query caching policy being attached when building a Lucene Searcher):
https://github.com/elastic/elasticsearch/blob/543588e7475fde86e77744b599ca89aa62dc6f45/server/src/main/java/org/elasticsearch/index/engine/Engine.java#L1068-L1075

Maybe something like "InternalEngine#getIndexWriterConfig() assembles the main Lucene write-path customizations for the `IndexWriter`, including the deletion policy, soft-deletes field, merge-policy wrapping, and codec selection."

Author: Thanks, updated this section.
